Support MSC4140: Delayed events (Futures)

AndrewFerr opened the pull request:
See: [MSC4140](https://github.com/matrix-org/matrix-spec-proposals/pull/4140)

### Pull Request Checklist

<!-- Please read https://element-hq.github.io/synapse/latest/development/contributing_guide.html before submitting your pull request -->

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

anoadragon453 commented on changelog.d/17326.feature:
```diff
@@ -0,0 +1,2 @@
+Add initial implementation of delayed events as proposed by [MSC4140](https://github.com/matrix-org/matrix-spec-proposals/pull/4140).
```
```suggestion
Add experimental implementation of delayed events as proposed by [MSC4140](https://github.com/matrix-org/matrix-spec-proposals/pull/4140).
```

anoadragon453 reviewed (changes requested):
@AndrewFerr would it be possible to break this PR up into a few more logical commits? Currently it's roughly 1 commit with >1000 lines changed, which appears daunting to review :slightly_smiling_face:

AndrewFerr commented:
Ironically, the reason I had squashed the commits was to try make it easier to review :upside_down_face: 

I've now split them up to isolate what should be the most relevant components of the PR.

anoadragon453 reviewed (changes requested):
Reviewed up until https://github.com/element-hq/synapse/commit/54725883c602850f674ec482ab28155b380fab9c, hopefully this gets you started. I'm afraid I'm away for the next few days, so requesting review from the rest of the team.

AndrewFerr commented on synapse/storage/schema/main/delta/87/01_add_delayed_events.sql:
```diff
@@ -0,0 +1,47 @@
+-- <https://www.gnu.org/licenses/agpl-3.0.html>.
+
+CREATE TABLE delayed_events (
+    delay_rowid INTEGER PRIMARY KEY, -- An alias of rowid in SQLite
+    delay_id TEXT NOT NULL,
```
This was done to have a column that is a "pure" primary key.  Currently, `delay_id`s are non-unique across users (hence the `UNIQUE (delay_id, user_localpart)` constraint) to allow for a greater range of possible `delay_id`s per user, and to scope `delay_id` lookups to the calling user (so that even if one were to be stolen/guessed by another user, they would get a 404 when trying to use it).  This wouldn't be possible the `delay_id` alone were to be the primary key.

And it's not so much that `delay_rowid` matches `rowid` in SQLite, it's that it _is_ the `rowid`: all SQLite tables implicitly have an integer key to uniquely identify rows, and giving a table a `INTEGER PRIMARY KEY` column makes it a named alias of that key.  Best of all, as long as `INSERT`s don't mention that column (or set it to `NULL`), it gets set to an auto-generated value that's guaranteed to be unique (which is even better than an auto-incrementing column, which has to be managed for overflows).  Put together, this lets `delay_rowid` be a DB-managed primary key.

For more info, see https://www.sqlite.org/lang_createtable.html#rowid. I will add that link in the comment.

AndrewFerr replied:
fafaa03214

AndrewFerr commented on synapse/storage/schema/main/delta/87/01_add_delayed_events.sql.postgres:
```diff
@@ -0,0 +1,14 @@
+-- See the GNU Affero General Public License for more details:
+-- <https://www.gnu.org/licenses/agpl-3.0.html>.
+
+ALTER TABLE delayed_events ALTER COLUMN delay_rowid ADD GENERATED ALWAYS AS IDENTITY;
```
Sure. It's similar to SQLite's `rowid`, but has to be opted into.

AndrewFerr replied:
fafaa03214

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,227 @@
+        """
+        Creates a new delayed event.
+
+        Params:
```
Ahh, this was bugging me!  Happy to learn that there's a style guide for this.

AndrewFerr replied:
b3d4d6cb41

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,227 @@
+
+class DelayedEventsHandler:
+    def __init__(self, hs: "HomeServer"):
+        self.store = hs.get_datastores().main
+        self.config = hs.config
+        self.clock = hs.get_clock()
+        self.request_ratelimiter = hs.get_request_ratelimiter()
+        self.event_creation_handler = hs.get_event_creation_handler()
+        self.room_member_handler = hs.get_room_member_handler()
```
Other handlers had seemingly-private attributes that weren't `_`-prefixed, so I wasn't sure what to do here.  These are definitely not meant to be used anywhere else, so I will go ahead and rename these.

AndrewFerr replied:
3afa3cf8f2

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,227 @@
+            content: The content of the event to be sent.
+            delay: How long (in milliseconds) to wait before automatically sending the event.
+                If None, the event won't be automatically sent (allowed only when parent_id is set).
+            parent_id: The ID of the delayed event this one is grouped with.
+                May only refer to a delayed event that has no parent itself.
```
No it isn't; I'll add a comment.

AndrewFerr replied:
48d81262b5

AndrewFerr commented on synapse/rest/client/room.py:
```diff
@@ -289,6 +290,24 @@ async def on_PUT(
             origin_server_ts = parse_integer(request, "ts")
 
+        delay = parse_integer(request, "org.matrix.msc4140.delay")
+        parent_id = parse_string(request, "org.matrix.msc4140.parent_delay_id")
```
Sure.  I included them to avoid having to bump the DB schema again if they were to be included later, and to work with clients that support them; but if getting the parent behaviour reviewed will slow down this PR, I'll strip them out.

AndrewFerr replied:
[`21311fb`](https://github.com/element-hq/synapse/pull/17326/commits/21311fbcb30a4b76876faa71465db05003f4d2ea)

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,227 @@
+        state_key: Optional[str],
+        origin_server_ts: Optional[int],
+        content: JsonDict,
+        delay: Optional[int],
+        parent_id: Optional[str],
```
Yes, delays & parent relations are compatible with each other.

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,227 @@
+
+        await self.request_ratelimiter.ratelimit(requester)
+
+        # TODO: Validate that the event is valid before scheduling it
```
They do get validated at send time, but this comment refers to checking that the event content itself isn't malformed, lest someone schedules an event that is doomed to fail no matter what.  This is handled in 32cbacf4c8.

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,358 @@
+
+            return delay_id
+
+        attempts_remaining = 10
+        while True:
+            try:
+                return await self.db_pool.runInteraction("add", add_txn)
+            except SynapseError as e:
+                if (
+                    e.code == HTTPStatus.INTERNAL_SERVER_ERROR
+                    and attempts_remaining > 0
+                ):
+                    attempts_remaining -= 1
+                else:
+                    raise e
```
This is for handling the unlikely case of generating a `delay_id` that matches an already-existing one & getting a collision.  The intent is to regenerate the ID until a non-colliding one is made (though it should be incredibly rare for a collision to ever happen).  If there's a better pattern for handling this, I'm all ears.

AndrewFerr replied:
90cc8b5

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,358 @@
+    random_string = stringutils.random_string(20)
+    base = f"syd_{random_string}"
+
+    crc = base62_encode(crc32(base.encode("ascii")), minwidth=6)
```
I don't think it's strictly needed, but other ID-like values we have use a CRC so I figured I'd use it here too.  The delay ID just needs to be an arbitrary, opaque string, so maybe it shouldn't bother with using a CRC.

AndrewFerr replied:
221e0af65f

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,437 @@
+
+        try:
+            if state_key is not None and event_type == EventTypes.Member:
+                membership = content.get("membership", None)
```
Turns out this is done in a few places (which I must have copied from :facepalm: ).  I'll probably make a small, separate PR for updating all of them.

AndrewFerr replied:
Interesting...using `, None)` makes mypy deduce `membership` as an `Any` instead of an `Optional[Any]`, so it actually worsens typechecking!

AndrewFerr replied:
56abbb95d7 just for this one

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,589 @@
+        Returns: The generated ID assigned to the added delayed event.
+
+        Raises:
+            SynapseError if the delayed event failed to be added.
```
Sure, but many other `Raises:` sections don't put a colon after the exception types they mention.  Maybe they should all be updated.

AndrewFerr replied:
c4e80adcbe, just for the ones in this PR

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,589 @@
+                - The IDs of all removed delayed events with a timeout that must be unscheduled.
+
+        Raises:
+            NotFoundError if there is no matching delayed event.
```
Ditto

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,358 @@
+                )
+
+            if not self._use_returning:
+                txn.execute(
+                    """
+                    SELECT delay_rowid
+                    FROM delayed_events
+                    WHERE delay_id = ? AND user_localpart = ?
+                    """,
+                    (
+                        delay_id,
+                        user_localpart,
+                    ),
```
Good idea -- `not self._use_returning` also means that SQLite3 is in use, which supports `last_insert_rowid`.

AndrewFerr replied:
Actually, no -- I had considered using it before, but found that it returns not the last rowID of a target table, but the last one among _all_ tables.  That means if some other table outside of this transaction had a row added to it immediately after the `delayed_events` table's row was inserted, the new rowID of that other table would be returned here instead.

AndrewFerr replied:
Oh, that's perfect then!

AndrewFerr replied:
This query is gone now, but if something like this is needed again, I'll remember to use `last_insert_rowid` for it.

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,358 @@
+
+        parent_rowid = self.db_pool.simple_select_one_onecol_txn(
+            txn,
+            table="delayed_event_children JOIN delayed_events ON child_rowid = delay_rowid",
```
Yes, the method's implementation indeed allows for selecting from joined tables.  If this is a code smell, I don't mind changing it, but there are some handy checks in the `simple_select_`*s that make them desirable over working with `execute`.

AndrewFerr replied:
This is gone now that parent IDs are removed by 21311fb.

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,437 @@
+        user_localpart: UserLocalpart,
+        delay: Delay,
+    ) -> None:
+        """NOTE: Should not be called with a delay_id that isn't in the DB, or with a negative delay."""
```
Checking that it's in the DB requires an async call, but I want to keep this method synchronous.  It's no longer that bad for the delay to be missing from the DB, so I'll just remove this comment.

Asserting for a positive delay is easy & a good idea, so it shall be done.

AndrewFerr replied:
08f54ca54c

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,589 @@
+            WITH target_rowids AS (
+                SELECT delay_rowid
+                FROM delayed_events
+                WHERE {" AND ".join("%s = ?" % k for k in keyvalues)}
```
It's tricky to use it here since the items to match on aren't known right away / require a DB lookup.

Plus, IMO generating a long `IN (?,?,...?)` clause may be less efficient than using a CTE, which lets the DB engine handle the complexity of deciding what to match on.

Also, this uses the same construction of the "WHERE...AND" clause used by `DatabasePool` methods (albeit using an f-string instead of `%`-formatting), so this should be as resistant to SQL injections as they are.

AndrewFerr replied:
This clause is gone now that parent IDs are removed by 21311fbcb3.

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,227 @@
+                )
+                event_id = event.event_id
+        except ShadowBanError:
+            event_id = "$" + random_string(43)
```
bee52bd95f

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,358 @@
+                        "Invalid parent delayed event",
+                        Codes.INVALID_PARAM,
+                        # TODO: Maybe remove this
+                        additional_fields={"db_error": str(e)},
```
62527082e6 moves the full error string to a debug message (to help debug what the DB error is when/if they happen).

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,358 @@
+                EventType(event_row[1]),
+                StateKey(event_row[2]) if event_row[2] is not None else None,
+                Timestamp(event_row[3]) if event_row[3] is not None else None,
+                # TODO: Verify contents?
```
The contents are now verified both on event scheduling time & send time, so this TODO is no longer needed: b1f74a81c3

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,358 @@
+        hs: "HomeServer",
+    ):
+        super().__init__(database, db_conn, hs)
+        # TODO: Always use RETURNING once the minimum supported Sqlite version is 3.35.0
```
https://github.com/element-hq/synapse/issues/17577

AndrewFerr commented:
I've gone through all of @anoadragon453's comments so far, so anyone else can feel free to pick this up :slightly_smiling_face:

andybalaam commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,227 @@
+# See the GNU Affero General Public License for more details:
+# <https://www.gnu.org/licenses/agpl-3.0.html>.
+#
+# Originally licensed under the Apache License, Version 2.0:
```
Is this needed for new files?

AndrewFerr replied:
I'm not sure, but given that other recently-added files have it, I assume they do.

ara4n replied:
we should only have the "originally licensed" for files which /were/ originally licensed as apache (i.e. not this one)

AndrewFerr replied:
Thanks. Removed them with 798c79e773

sandhose commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,461 @@
+    """Generates an opaque string, for use as a delay ID"""
+
+    # We use the following format for delay IDs:
+    #    syf_<random string>
```
```suggestion
    #    syd_<random string>
```

AndrewFerr replied:
Oops, thanks! 347811891a

sandhose commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,461 @@
+                )
+                if row is None:
+                    raise NotFoundError("Delayed event not found")
+                self.db_pool.simple_update_txn(
+                    txn,
+                    table="delayed_events",
+                    keyvalues={"delay_rowid": row[0]},
+                    updatevalues={"running_since": current_ts},
+                )
+                assert txn.rowcount == 1
```
```suggestion
                self.db_pool.simple_update_one_txn(
                    txn,
                    table="delayed_events",
                    keyvalues={"delay_rowid": row[0]},
                    updatevalues={"running_since": current_ts},
                )
```

AndrewFerr replied:
e0e6802f1d

sandhose commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,461 @@
+                assert txn.rowcount == 1
+            return Delay(row[1])
+
+        return await self.db_pool.runInteraction("restart", restart_txn)
```
The interaction name should be more descriptive, as I believe it bubbles up in metrics and traces

AndrewFerr replied:
85cb72f9f6

sandhose commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,418 @@
+
+        async def _schedule_db_events() -> None:
+            # TODO: Sync all state first, so that affected delayed state events will be cancelled
+            events, remaining_timeout_delays = await self._store.process_all_delays(
+                self._get_current_ts()
+            )
```
I'm slightly scared that this removes them from the delay table, only living in-process. If the process crashes in the meantime, they would get lost

AndrewFerr replied:
This runs only once on startup, at which point there will be no in-process timers yet.

AndrewFerr replied:
be094e6 avoids removing from the delay table until the event has persisted.

sandhose commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,418 @@
+            SynapseError: if the provided action is unknown, or is unsupported for the target delayed event.
+            NotFoundError: if no matching delayed event could be found.
+        """
+        try:
+            enum_action = _UpdateDelayedEventAction(action)
+        except ValueError:
+            raise SynapseError(
+                HTTPStatus.BAD_REQUEST,
+                "'action' is not one of "
+                + ", ".join(f"'{m.value}'" for m in _UpdateDelayedEventAction),
+                Codes.INVALID_PARAM,
+            )
```
Feels like this should be checked at the REST API level?

AndrewFerr replied:
8a65c77336

sandhose commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,418 @@
+        Returns:
+            The ID of the added delayed event.
+        """
+        max_delay = self._config.experimental.msc4140_max_delay
+        if delay > max_delay:
+            raise SynapseError(
+                HTTPStatus.BAD_REQUEST,
+                "The requested delay exceeds the allowed maximum.",
+                Codes.UNKNOWN,
+                {
+                    "org.matrix.msc4140.errcode": "M_MAX_DELAY_EXCEEDED",
+                    "org.matrix.msc4140.max_delay": max_delay,
+                },
+            )
```
Feels like this should be checked at the REST API level?

AndrewFerr replied:
47b6e69

sandhose commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,418 @@
+                self._schedule(delay_id, user_localpart, delay)
+
+            elif enum_action == _UpdateDelayedEventAction.SEND:
+                args = await self._store.pop_event(delay_id, user_localpart)
+
+                self._unschedule(delay_id, user_localpart)
+                await self._send_event(user_localpart, *args)
```
This also means the even could get lost here

AndrewFerr replied:
be094e6

sandhose commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,418 @@
+    ) -> None:
+        user_id = UserID(user_localpart, self._config.server.server_name)
+        user_id_str = user_id.to_string()
+        requester = create_requester(
+            user_id,
+            is_guest=await self._store.is_guest(user_id_str),
+        )
```
This feels a bit clunky, but I understand why it's like that :(
We should *at least* keep the `device_id` in the requester as well, because we use that to scope transaction IDs and handle local echo

Can you add a comment here about why we're creating a requester on the fly here?

AndrewFerr replied:
This excludes a `device_id` because the device that scheduled the delayed event isn't really involved in the eventual sending of that event, or so I thought.

Should the DB save (along with the event content to send) the device ID that requested the delayed event?

AndrewFerr replied:
Something else to consider is that the device (and its access token) which requested a delayed event may be destroyed / logged out by the time the event actually gets sent.

If the device ID / access token really is needed for this, how about adding a DB column to store the requester of the delayed event?  I had tried that before (thanks to requesters being serializable) but got rid of it due to thinking devices were irrelevant for delayed event sending.

AndrewFerr replied:
235c432 adds the device ID and the comment.

sandhose commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,461 @@
+        attempts_remaining = 10
+        while True:
+            try:
+                return await self.db_pool.runInteraction("add", add_txn)
```
Same here, this should use a more descriptive interaction name

AndrewFerr replied:
85cb72f9f6

sandhose commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,461 @@
+                else:
+                    raise e
+
+    async def restart(
```
```suggestion
    async def restart_delayed_event(
```

AndrewFerr replied:
85cb72f9f6

sandhose commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,461 @@
+
+# TODO: Try to support workers
+class DelayedEventsStore(SQLBaseStore):
+    async def add(
```
```suggestion
    async def add_delayed_event(
```

AndrewFerr replied:
85cb72f9f6

sandhose commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,461 @@
+
+        return await self.db_pool.runInteraction("restart", restart_txn)
+
+    async def get_all_for_user(
```
```suggestion
    async def get_all_delayed_events_for_user(
```

AndrewFerr replied:
85cb72f9f6

sandhose commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,461 @@
+            for row in rows
+        ]
+
+    async def process_all_delays(self, current_ts: Timestamp) -> Tuple[
```
```suggestion
    async def process_all_delayed_events(self, current_ts: Timestamp) -> Tuple[
```

sandhose commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,461 @@
+            "process_all_delays", process_all_delays_txn
+        )
+
+    async def pop_event(
```
```suggestion
    async def pop_delayed_event(
```

AndrewFerr replied:
85cb72f9f6

sandhose commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,461 @@
+
+        return await self.db_pool.runInteraction("pop_event", pop_event_txn)
+
+    async def remove(
```
```suggestion
    async def remove_delayed_event(
```

AndrewFerr replied:
85cb72f9f6

sandhose commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,461 @@
+            desc="remove",
+        )
+
+    async def remove_state_events(
```
```suggestion
    async def remove_delayed_state_events(
```

AndrewFerr replied:
85cb72f9f6

sandhose commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,418 @@
+
+            await self._send_event(user_localpart, *args)
+
+    def _schedule(
```
I don't like the idea of having all the delayed events as in-process timeouts like that. I would rather have a reconciliation loop, where it fetches the next deadline, and wakes up at that time to process any pending event. This would greatly simplify the logic here

AndrewFerr replied:
be094e6

sandhose commented on synapse/storage/schema/main/delta/87/01_add_delayed_events.sql:
```diff
@@ -0,0 +1,34 @@
+    event_type TEXT NOT NULL,
+    state_key TEXT,
+    origin_server_ts BIGINT,
+    content bytea NOT NULL,
```
One way to avoid loosing event could be to:

 - add a column like `processed_ts`, or `picked_up_at` (I'm not good at naming things) where we record when the event is picked up by a worker
 - record the delay ID in the event internal metadata
 - when an event gets properly persisted, delete it from the delayed events
 - if we find events that have a `processed_ts` too old (5 min or so), consider that they were not persisted, clear that column and they would get processed again

AndrewFerr replied:
Thanks for thinking so much about lost events, it's been a huge help!

Another situation that arises is the risk of double-sending an event: if the server crashes after a delayed event is persisted but before it's removed from the DB table of pending delayed events, a server restart would have to know whether an entry of the delayed event table is one that actually got persisted & doesn't need to be sent again.

I discussed this with the VoIP team, and a strategy we arrived at was to search through the timeline up to a delayed event's `send_ts` for events possibly sent during a server outage, and if there's one that has that event's delay ID in its metadata (either in the `"unsigned"` object or in some Synapse-internal struct/table), then it means the delayed event was sent & doesn't need to be sent again.  This avoids needing the `processed_ts` column, because we know that only events in the past would have been processed, and we already have the timestamp of them via `send_ts`.

Though what would be ideal is if event persistence could include a step for deleting delayed events from the DB, as long as the work done for event persistence is atomic / done in a transaction.  Then there'd be a guarantee that the delayed event table would never get out-of-sync with the set of persisted events.

AndrewFerr replied:
be094e6, but still missing is a scan of the timeline to avoid double-sending events that persisted but weren't removed from the delay table.  That is hopefully a rare enough case to not be a showstopper, though.

AndrewFerr replied:
I'm going to leave the double-sending fix to another PR.

sandhose commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,418 @@
+            )
+
+        await self._request_ratelimiter.ratelimit(requester)
+        await self._initialized_from_db
```
If I get that correctly, that means that we'd block until all the events in the queue are persisted?
So if your server is away for a while, and all the sudden you have thousands of delayed events to process, any `add`/`update` call would block until they get processed, instead of getting queued

AndrewFerr replied:
This should wait only on delayed events that were loaded from the DB on startup, i.e. delayed events that were scheduled (and not yet sent) prior to a server restart.

Since `initialized_from_db` is the `Deferred` returned by the async work done on startup, AIUI `await`ing it doesn't repeat any work, and instead just checks that the startup work is done before continuing.

The waits that are relevant to newly-added delayed events are the ones from `_get_delay_context` and `_get_state_context`, which should be scoped to `delay_id` and `state_key` respectively. (But now that I look at it, `_get_state_context` should also be scoped to rooms).

AndrewFerr replied:
The startup task now blocks only on the job of marking timed out events as being processed, but the actual sending of the events is done in the background.

AndrewFerr commented on synapse/config/experimental.py:
```diff
@@ -443,6 +443,20 @@ def read_config(self, config: JsonDict, **kwargs: Any) -> None:
+                "msc4140_max_delay must be a positive integer",
+                ("experimental", "msc4140_max_delay"),
+            )
+        except KeyError:
+            self.msc4140_max_delay = 10 * 365 * 24 * 60 * 60 * 1000  # 10 years
```
Skipping the check is solely for making it more clear that the default value will never throw a `ValueError`.

00217f3152 reraises the `ValueError`; if it's still preferred to drop the `KeyError`, I'll do it.

AndrewFerr replied:
I refactored this to use similar code style as other duration configs: 3ce7305

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,461 @@
+                ):
+                    attempts_remaining -= 1
+                else:
+                    raise e
```
Removed by 90cc8b5a22

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,461 @@
+                raise SynapseError(
+                    HTTPStatus.INTERNAL_SERVER_ERROR,
+                    f"Couldn't generate a unique delay_id for user_localpart {user_localpart}",
+                )
```
Removed by 90cc8b5a22

AndrewFerr commented on synapse/storage/schema/main/delta/87/01_add_delayed_events.sql:
```diff
@@ -0,0 +1,34 @@
+    -- Newly-inserted rows that don't assign a (non-NULL) value for this column
+    -- will have it set to a table-unique value.
+    -- For Postgres to do this, the column must be set as an identity column.
+    delay_rowid INTEGER PRIMARY KEY,
```
The main benefit of the single-column rowid key was to make joins on other delayed-event tables easier, but those are gone now, so this can be done: 5672d0d973

AndrewFerr replied:
Also 05accda65b

AndrewFerr commented on synapse/storage/schema/main/delta/87/01_add_delayed_events.sql:
```diff
@@ -0,0 +1,34 @@
+    delay_id TEXT NOT NULL,
+    user_localpart TEXT NOT NULL,
+    delay BIGINT NOT NULL,
+    running_since BIGINT NOT NULL,
```
Turns out it's possible to index an expression like `running_since + delay`, but having the send time as a column is more convenient, so I've done it: 57b7229

Next is using it to fetch the nearest deadline.

AndrewFerr replied:
be094e6

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,358 @@
+]
+
+
+# TODO: Try to support workers
```
Workers should be better supported now, though other comments address open questions.

AndrewFerr commented on synapse/config/server.py:
```diff
@@ -780,6 +780,17 @@ def read_config(self, config: JsonDict, **kwargs: Any) -> None:
         else:
             self.delete_stale_devices_after = None
 
+        # The maximum allowed delay duration for delayed events (MSC4140).
+        max_event_delay_duration = config.get("max_event_delay_duration")
+        if max_event_delay_duration is not None:
+            self.max_event_delay_ms: Optional[int] = self.parse_duration(
+                max_event_delay_duration
+            )
+            if self.max_event_delay_ms <= 0:
+                raise ConfigError("max_event_delay_duration must be a positive value")
+        else:
+            self.max_event_delay_ms = None
+
```
This config no longer in the `experimental` section, but is now disabled by default.  The reasoning is to reduce config churn once delayed events become stable, while still making it something server admins must explicitly opt into using.

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,414 @@
+
+        self._next_delayed_event_call: Optional[IDelayedCall] = None
+
+        # TODO: Looks like these callbacks are run in background. Find a foreground one
+        hs.get_module_api().register_third_party_rules_callbacks(
+            on_new_event=self.on_new_event,
+        )
```
The aim is to cancel all delayed state events whenever a conflicting state event (i.e. one with the same room ID + event type + state key) is received.

The problem is that the only callback I could find for reacting to received events is the third party callbacks, and those are called as a background task during event persistence: https://github.com/element-hq/synapse/blob/f4032d3e718c806a193fb5167d9958052cd47add/synapse/handlers/message.py#L1935

Is there anything that runs a foreground callback when an event is received?  And how do workers get notified of newly received events?

AndrewFerr replied:
> Module API callback functions should only be registered by [Synapse modules](https://element-hq.github.io/synapse/latest/modules/index.html),

Ah, I was wondering about that. I found it used by the [InviteAutoAccepter](https://github.com/element-hq/synapse/blob/e5d07bb0830eea056ea85863bca3d5b093d43a88/synapse/events/auto_accept_invites.py#L57) so I thought using it for internal callbacks was acceptable, but it makes more sense that it's meant for external modules.

> In this case, likely what you want is to add a replication callback for the event notifier (see [example](https://github.com/element-hq/synapse/blob/ca69d0f57165ecb10204ee433992b20af71cbe91/synapse/handlers/presence.py#L868-L869)).

Ideally, I need a callback that is told about newly-received events (namely state events), but the replication callback isn't given any arguments. Is there a callback for received events?  If not, should I create one?  It would be analogous to [the callback for joins](https://github.com/element-hq/synapse/blob/e5d07bb0830eea056ea85863bca3d5b093d43a88/synapse/notifier.py#L297).

AndrewFerr replied:
> Module API callback functions should only be registered by [Synapse modules](https://element-hq.github.io/synapse/latest/modules/index.html),

Ah, I was wondering about that. I found it used by the [InviteAutoAccepter](https://github.com/element-hq/synapse/blob/e5d07bb0830eea056ea85863bca3d5b093d43a88/synapse/events/auto_accept_invites.py#L57) so I thought using it for internal callbacks was acceptable, but it makes more sense that it's meant for external modules.

> In this case, likely what you want is to add a replication callback for the event notifier (see [example](https://github.com/element-hq/synapse/blob/ca69d0f57165ecb10204ee433992b20af71cbe91/synapse/handlers/presence.py#L868-L869)).

Ideally, I need a callback that is told about newly-received events (namely state events), but the replication callback isn't given any arguments. Is there a callback for received events?  If not, should I create one?  It would be analogous to [the callback for joins](https://github.com/element-hq/synapse/blob/e5d07bb0830eea056ea85863bca3d5b093d43a88/synapse/notifier.py#L297).

anoadragon453 replied:
> Ah, I was wondering about that. I found it used by the [InviteAutoAccepter](https://github.com/element-hq/synapse/blob/e5d07bb0830eea056ea85863bca3d5b093d43a88/synapse/events/auto_accept_invites.py#L57) so I thought using it for internal callbacks was acceptable, but it makes more sense that it's meant for external modules.

Yes, this was an external module that was inlined into the codebase. There was [a back and forth](https://github.com/element-hq/synapse/pull/17147#discussion_r1592629098) about using external module APIs internally, though ultimately we ended up leaving it how it was.

> Ideally, I need a callback that is told about newly-received events (namely state events), but the replication callback isn't given any arguments. Is there a callback for received events?  If not, should I create one?  It would be analogous to [the callback for joins](https://github.com/element-hq/synapse/blob/e5d07bb0830eea056ea85863bca3d5b093d43a88/synapse/notifier.py#L297).

Ah, right, sorry. The way this works is not that events are sent over replication, but rather workers are merely poked that new events exist, and are then expected on their own to go and pull them out of the database. This cuts down on the size of the replication traffic, but does mean you need to do extra work.

Synapse has the concept of [streams](https://element-hq.github.io/synapse/latest/development/synapse_architecture/streams.html). Each event in the `events` table has a "stream ID" (`stream_ordering` column). Workers have a stream ID they keep track of, and upon being told that there are new events, will query the database for "all events with X attributes that have a stream ID > Y" to get at new events.

If you're just after state event changes, you can make use of the `current_state_deltas_stream`. It only track state events, rather than all event types.

The presence example might not be best as it always just [starts from the latest event on startup](https://github.com/element-hq/synapse/blob/ca69d0f57165ecb10204ee433992b20af71cbe91/synapse/handlers/presence.py#L871-L873). A better mechanism to reference is the [`UserDirectoryHandler`](https://github.com/element-hq/synapse/blob/anoa/configure_workers_updates/build/lib.linux-x86_64-cpython-312/synapse/handlers/user_directory.py#L71). It stores its currently read up to stream token in the [`user_directory_stream_pos` table](https://github.com/element-hq/synapse/blob/anoa/configure_workers_updates/build/lib.linux-x86_64-cpython-312/synapse/handlers/user_directory.py#L209-L211), which is queried on startup. [`notify_new_event`](https://github.com/element-hq/synapse/blob/anoa/configure_workers_updates/build/lib.linux-x86_64-cpython-312/synapse/handlers/user_directory.py#L171-L186) is then called whenever a new event is created, and then the `current_state_deltas_stream` is read to see whether users have left/joined any rooms. The user directory is then updated accordingly.

It's possible to have multiple workers reading here, but it's easiest to start with just one.

AndrewFerr replied:
This was crucial info, thanks for such a helpful writeup!  [`d36c89f`](https://github.com/element-hq/synapse/pull/17326/commits/d36c89f10e2d87c497fbd54533c0d7e421024cfa) handles this.

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,414 @@
+            # - This will interfere with workers that are in the act of processing delayed events.
+            # TODO: To avoid double-sending, scan the timeline to find which of these events were
+            # already sent. To do so, must store delay_ids in sent events to retrieve them later.
+            # TODO: To avoid interfering with workers, think of a way to distinguish between
+            # events being processed by a worker vs ones that got lost after a server crash.
```
What `unprocess_delayed_events` is trying to solve is the case of the server crashing between the time it marked delayed events as `is_processed` but before having been able to send & persist the event.  That would leave behind `in_processed` rows the next time the server starts, but with nothing running to handle sending the events.  So, the quick fix is to try sending them again, by unmarking them as processed and treating them like any other timed-out delayed event.

The problem is that, when using workers, `is_processed` rows being present on startup could also mean that another worker is in the middle of sending delayed events.

Maybe the right approach is to allow only one worker / the main process to handle sending delayed events?  Currently, the timer for sending delayed events can get set for multiple workers, but the way `is_processed` is set & used for filtering rows should prevent multiple workers from trying to send/edit the same event at once.  But it doesn't help with this startup interference issue.

AndrewFerr replied:
Restricting delayed event sending to a single worker should definitely simplify things, but I still have to look into how to handle send-on-timeout vs send-on-demand.  With the way things are, multiple workers may already be safe, but maybe not...

anoadragon453 replied:
How is this looking now after discussing how to use the background worker? Can we simplify the mechanism at all?

anoadragon453 replied:
Alas, being able to have `/send` and `/state` on multiple event creator workers is currently a big part of Synapse's horizontal scalability capabilities.

Having multiple workers adding events to delayed state tables, and only a single worker pulling them out/cancelling them could work?

AndrewFerr replied:
092793a287 restricts as much as possible to a single worker.

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,414 @@
+            # - This will double-send delayed events that successfully persisted, but failed to be
+            #   removed from the DB table of delayed events.
+            # - This will interfere with workers that are in the act of processing delayed events.
+            # TODO: To avoid double-sending, scan the timeline to find which of these events were
+            # already sent. To do so, must store delay_ids in sent events to retrieve them later.
```
To recap https://github.com/element-hq/synapse/pull/17326#discussion_r1729227576 & https://github.com/element-hq/synapse/pull/17326#discussion_r1734343222, the solution I had in mind for this was to:
- put the delay ID of each sent delayed event in its `"unsigned"` section
- on startup, for each `is_processed` row still in the delayed events table, scan its destination room until its `send_ts`, searching for any event with its delay ID in the `"unsigned"` section.  If found, skip sending the delayed event; otherwise, do send it.

AndrewFerr replied:
I'm going to leave this to another PR.

anoadragon453 commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,529 @@
+                assert old_next_send_ts is not None
+                new_next_send_ts = self._get_next_delayed_event_send_ts_txn(txn)
+                return new_next_send_ts != old_next_send_ts, new_next_send_ts
+            else:
+                return False, None
```
nit

```suggestion
            
            return False, None
```

AndrewFerr replied:
c3ad95d

anoadragon453 reviewed (changes requested):
Apologies for the long wait time on this PR. It's quite a hefty one, with new concepts, and I kept getting pulled away before I could fully wrap my brain around it.

A fair number of comments below, but nothing requiring any major architecting. I would like to see some more testing done, either through unit tests or complement, to help prove the handling of sending the next delayed event.

It would also be helpful to include docstrings for private methods, even if they're small, as it helps when skimming the code while getting a grasp on it for the first time.

(Note: a few comments below may be outdated as they were made on commits. Instead of clicking through all the commits and closing them, I'm just going to do so after submitting this review.)

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -45,6 +46,12 @@
+class _UpdateDelayedEventAction(Enum):
+    CANCEL = "cancel"
+    RESTART = "restart"
+    SEND = "send"
```
Looks like `Enum` fields are already effectively `Final`.  If I try to overwrite one of these without hinting it as `Final`, mypy raises an error of "Cannot assign to final attribute".

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,529 @@
+        delay_id: str,
+        user_localpart: str,
+    ) -> Tuple[
+        Tuple[
+            RoomID,
+            EventType,
+            Optional[StateKey],
+            Timestamp,
+            JsonDict,
+            Optional[DeviceID],
+        ],
```
This excludes the `DelayID` and `UserLocalpart` fields.  I'd like for this to be a named type that `DelayedEventDetails` could extend from / include, but couldn't think of a clean way to do it.  Maybe defining both of them as `attr` classes is the answer.

AndrewFerr replied:
a60fa7f91b

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,396 @@
+                )
+                if state_info is not None:
+                    # Note that removal from the DB is done by self.on_new_event
+                    sent_state.add(state_info)
```
It's used a few lines up, to avoid sending multiple timed-out delayed events for the same state.

anoadragon453 replied:
Ah! Thanks, not sure how I missed that :yum:

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -0,0 +1,396 @@
+        user_id = UserID(user_localpart, self._config.server.server_name)
+        user_id_str = user_id.to_string()
+        # Create a new requester from what data is currently available
+        # TODO: Consider storing the requester in the DB at add time and deserialize it here
```
That's good to know, thanks!

Are there any other properties that should be sent to `create_requester` here?  If not, I'll just remove the comment and consider this sufficient.

anoadragon453 replied:
What you're passing now is sufficient I think.

AndrewFerr replied:
1dbbb749bc

AndrewFerr commented on synapse/rest/client/delayed_events.py:
```diff
@@ -0,0 +1,95 @@
+
+    async def on_GET(self, request: SynapseRequest) -> Tuple[int, List[JsonDict]]:
+        requester = await self.auth.get_user_by_req(request)
+        # TODO: Support Pagination stream API ("from" query parameter)
```
In a follow-up, and agreed.

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,529 @@
+
+            return old_next_send_ts is None or send_ts < old_next_send_ts
+
+        changed = await self.db_pool.runInteraction(
```
Yes, exactly

anoadragon453 replied:
In that case, I think the docstring's definition for the second item in the returned tuple is a little confusing. Rather than "whether the added delayed event is the next to be sent", perhaps it should be "whether the next event to be sent is now this event instead".

AndrewFerr replied:
1d750601, though I dropped the concept of checking for a changed next-send-time (at least for now) in a6cf11c0.

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,529 @@
+                },
+            )
+
+            return old_next_send_ts is None or send_ts < old_next_send_ts
```
Yes it does, but the only purpose of the return value is to know whether or not to reschedule the check for timed-out delayed events.  Whether the added event is the next to be sent or not doesn't matter, since the timeout check scans for _all_ timed-out events, meaning any events with the exact same timeout time will all get sent.

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,529 @@
+                    updatevalues={"send_ts": restarted_send_ts},
+                )
+
+            new_next_send_ts = self._get_next_delayed_event_send_ts_txn(txn)
```
Yes it can; I'll refactor the DB lookups to achieve this.

AndrewFerr replied:
Er, what I should have said was "not quite", because the current DB lookups aren't enough to deduce this.  Changing what's queried can get the required information with fewer total requests to the DB.

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,529 @@
+            user_localpart: The localpart of the delayed event's owner.
+            current_ts: The current time, which will be used to calculate the new send time.
+
+        Returns: Whether the matching delayed event would have been the next to be sent,
```
It's not; in fact, since restarting a delayed event can only delay it further, a restarted event can never become the next one to be sent.

The return value is concerned only with whether the next-soonest timeout time among all delayed events has changed by virtue of the target event being restarted.  With that said, the current wording doesn't reflect that, so I'll update it.

AndrewFerr replied:
Refactored and redocumented in 3e9f76f32d

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,529 @@
+            current_ts: The current time, which will be used to calculate the new send time.
+
+        Returns: Whether the matching delayed event would have been the next to be sent,
+            and if so, what the next soonest send time is, if any.
```
It currently is, but it need not be.  If the returned `bool` is False, the returned `Timestamp` has no meaning and should be ignored (and I should make it `None` to be explicit about that).

The reason is that the `Timestamp` is meaningful only when the next-soonest timeout time has changed, because that time is used to reschedule the timeout callback; if the time didn't change, there's no need to reschedule anything.

anoadragon453 replied:
That makes sense, thanks. I agree that making it `None` would be a clearer signal to callers to not use it.

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,529 @@
+        user_localpart: str,
+    ) -> List[JsonDict]:
+        """Returns all pending delayed events owned by the given user."""
+        # TODO: Store and return "transaction_id"
```
Presumably, but I doubt its usefulness & return it just to satisfy the current version of the MSC.  I propose against returning it [here](https://github.com/matrix-org/matrix-spec-proposals/pull/4140/files#r1747605229).

anoadragon453 replied:
Right. We'll keep it for now while it matches the MSC then, but open to changing it down the line.

AndrewFerr replied:
Returning transaction IDs has been removed from the MSC, so dd3c746eb5 removes this TODO.

AndrewFerr commented on docs/usage/configuration/config_documentation.md:
```diff
@@ -761,6 +761,18 @@ email:
+The maximum allowed duration by which sent events can be delayed, as per MSC4140.
+Must be a positive value if set.
+
+Defaults to no duration, which disallows sending delayed events.
```
a3fbdd348d

AndrewFerr commented on docs/usage/configuration/config_documentation.md:
```diff
@@ -761,6 +761,18 @@ email:
+---
+### `max_event_delay_duration`
+
+The maximum allowed duration by which sent events can be delayed, as per MSC4140.
```
a3fbdd348d

AndrewFerr commented on synapse/rest/client/delayed_events.py:
```diff
@@ -0,0 +1,95 @@
+"""This module contains REST servlets to do with delayed events: /delayed_events/<paths>"""
```
ef7284ff3d

AndrewFerr commented on tests/rest/client/test_rooms.py:
```diff
@@ -2291,6 +2291,78 @@ def test_room_message_filter_wildcard(self) -> None:
+        self.room_id = self.helper.create_room_as(self.user_id)
+
+    def test_send_delayed_invalid_event(self) -> None:
+        """Test sending a delayed event with invalid content."""
```
Yes, good catch.  Fixed in 0ab82f50b5

AndrewFerr commented on tests/rest/client/test_rooms.py:
```diff
@@ -2291,6 +2291,78 @@ def test_room_message_filter_wildcard(self) -> None:
         self.assertEqual(len(chunk), 2, [event["content"] for event in chunk])
 
 
+class RoomDelayedEventTestCase(RoomBase):
```
Good idea.  This is done in 0ab82f50b5

AndrewFerr commented on tests/rest/client/test_rooms.py:
```diff
@@ -2291,6 +2291,78 @@ def test_room_message_filter_wildcard(self) -> None:
+
+    @unittest.override_config({"max_event_delay_duration": "24h"})
+    def test_send_delayed_message_event(self) -> None:
+        """Test sending a delayed event with invalid content."""
```
0ab82f50b5

AndrewFerr commented on tests/rest/client/test_rooms.py:
```diff
@@ -2291,6 +2291,78 @@ def test_room_message_filter_wildcard(self) -> None:
+
+    @unittest.override_config({"max_event_delay_duration": "24h"})
+    def test_send_delayed_state_event(self) -> None:
+        """Test sending a delayed event with invalid content."""
```
0ab82f50b5

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,529 @@
+        ) -> Tuple[bool, Timestamp]:
+            old_next_send_ts = self._get_next_delayed_event_send_ts_txn(txn)
+            if old_next_send_ts is None:
+                raise NotFoundError("Delayed event not found")
```
90259228ee (though it's not just for the requesting user, but any user)

AndrewFerr commented on synapse/storage/schema/main/delta/88/01_add_delayed_events.sql:
```diff
@@ -16,3 +16,13 @@ CREATE TABLE delayed_events (
 CREATE INDEX delayed_events_is_processed ON delayed_events (is_processed);
 CREATE INDEX delayed_events_room_state_event_idx ON delayed_events (room_id, event_type, state_key) WHERE state_key IS NOT NULL;
+
+CREATE TABLE delayed_events_stream_pos (
+    Lock CHAR(1) NOT NULL DEFAULT 'X' UNIQUE,  -- Makes sure this table only has one row.
+    stream_id BIGINT NOT NULL,
+    CHECK (Lock='X')
+);
+
+INSERT INTO delayed_events_stream_pos (
+    stream_id
+) SELECT COALESCE(MAX(stream_ordering), 0) from events;
```
This was based on [04_add_room_forgetter.sql](https://github.com/element-hq/synapse/blob/develop/synapse/storage/schema/main/delta/76/04_add_room_forgetter.sql).

AndrewFerr commented on synapse/handlers/delayed_events.py:
`notify_new_event` & company was based on similar code from [`UserDirectoryHandler`](https://github.com/element-hq/synapse/blob/e1ed959a68f8039130be821c27e82e75b5d59e5f/synapse/handlers/user_directory.py#L171), [`PresenceHandler`](https://github.com/element-hq/synapse/blob/develop/synapse/handlers/presence.py#L1478), [`RoomForgetterHandler`](https://github.com/element-hq/synapse/blob/e1ed959a68f8039130be821c27e82e75b5d59e5f/synapse/handlers/room_member.py#L2163), and [`StatsHandler`](https://github.com/element-hq/synapse/blob/e1ed959a68f8039130be821c27e82e75b5d59e5f/synapse/handlers/stats.py#L78).

AndrewFerr commented on synapse/module_api/callbacks/third_party_event_rules_callbacks.py:
```diff
@@ -426,10 +426,7 @@ async def on_new_event(self, event_id: str) -> None:
-        event = await self.store.get_event(event_id, allow_none=True)
-        if not event:
-            logger.warning("Could not find event %s" % (event_id,))
-            return
```
I reverted this since it was needed only for my third-party callback, which is gone now.

AndrewFerr commented on tests/rest/client/test_rooms.py:
I reverted these since they had changed only because of my third-party callback, which is gone now.

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,529 @@
+            for row in rows
+        ]
+
+    async def process_timeout_delayed_events(
```
This function is used not only on the startup handling (which I'd like to remove, FTR) but also on _any_ timeout.  The idea is to check for all timed-out events instead of just one, in case there was some delay between the timeout time & when it actually runs; and also to be more resilient against failures (so if one delayed event failed to send, the next timeout will send the next event + all prior ones).

anoadragon453 commented on synapse/storage/schema/main/delta/88/01_add_delayed_events.sql:
```diff
@@ -16,3 +16,13 @@ CREATE TABLE delayed_events (
+    CHECK (Lock='X')
+);
+
+INSERT INTO delayed_events_stream_pos (
```
```suggestion
-- Start processing events from the point this migration was run, rather
-- than the beginning of time.
INSERT INTO delayed_events_stream_pos (
```

anoadragon453 commented on synapse/handlers/delayed_events.py:
```diff
@@ -83,19 +91,80 @@ async def _schedule_db_events() -> None:
-    async def on_new_event(
-        self, event: EventBase, _state_events: StateMap[EventBase]
-    ) -> None:
+    def notify_new_event(self) -> None:
```
It may be important to note here with a comment that this isn't the mechanism that actually sends delayed events. This only checks for state events that may cancel a pending delayed state event.

anoadragon453 reviewed (changes requested):
A few small changes here, but overall things are really shaping up.

Tests-wise, it'd be nice to see some testing of:

- A new state event cancelling a pending delayed state event.
- Whether pending delayed events get sent correctly on startup.

AndrewFerr commented on synapse/handlers/delayed_events.py:
```diff
@@ -83,19 +91,80 @@ async def _schedule_db_events() -> None:
+            self._event_pos = await self._store.get_delayed_events_stream_pos()
+            room_max_stream_ordering = self._store.get_room_max_stream_ordering()
+            if self._event_pos > room_max_stream_ordering:
+                # apparently, we've processed more events than exist in the database!
+                # this can happen if events are removed with history purge or similar.
+                logger.warning(
```
This was copied from a few other places:
- https://github.com/element-hq/synapse/blob/e1ed959a68f8039130be821c27e82e75b5d59e5f/synapse/handlers/user_directory.py#L221
- https://github.com/element-hq/synapse/blob/e1ed959a68f8039130be821c27e82e75b5d59e5f/synapse/handlers/room_member.py#L2186
- https://github.com/element-hq/synapse/blob/e1ed959a68f8039130be821c27e82e75b5d59e5f/synapse/handlers/stats.py#L101

So maybe let's keep it the same for the sake of consistency, and update all of them in another PR if appropriate.  (On that note, there's quite a lot of boilerplate for following the state event stream, so it'd be nice to commonize it somehow...)

anoadragon453 commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -99,68 +99,94 @@ async def restart_delayed_event(
+                    # The restarted event is the next to be sent
+                    delay_id == next_event[0]
+                    and user_localpart == next_event[1]
+                    # The next event to be sent is the only one to be sent at that time
```
Yep, that sounds good as well!

anoadragon453 commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -99,68 +99,94 @@ async def restart_delayed_event(
+                raise NotFoundError("Delayed event not found")
+
+            if changed:
+                if self.database_engine.supports_returning:
+                    row = txn.fetchone()
+                    assert row is not None
+                    restarted_send_ts = Timestamp(row[0])
+                else:
+                    restarted_send_ts = Timestamp(
+                        self.db_pool.simple_select_one_onecol_txn(
+                            txn,
+                            table="delayed_events",
+                            keyvalues={
+                                "delay_id": delay_id,
+                                "user_localpart": user_localpart,
+                                "is_processed": False,
+                            },
+                            retcol="send_ts",
+                        )
+                    )
```
Ah, right of course. Carry on then!

anoadragon453 commented on synapse/handlers/delayed_events.py:
```diff
@@ -83,19 +91,80 @@ async def _schedule_db_events() -> None:
-        Checks if a received event is a state event, and if so,
-        cancels any delayed events that target the same state.
+        Process current state deltas to cancel pending delayed events
+        that target the same state as any received state events.
```
Oh, yes. Sorry, I wasn't saying the comment was wrong - just that it felt a bit repetitive.

AndrewFerr commented:
See https://github.com/matrix-org/complement/pull/734 for Complement tests.  So far, the only test is for new state events cancelling pending delayed events; more tests are to come.

AndrewFerr commented:
https://github.com/matrix-org/complement/pull/734 now has a test for sending timed-out / restoring still-pending delayed events on server restart.

AndrewFerr commented on synapse/storage/databases/main/delayed_events.py:
```diff
@@ -0,0 +1,529 @@
+    Timestamp,
+    JsonDict,
+    Optional[DeviceID],
+]
```
a60fa7f91b

AndrewFerr commented on synapse/rest/client/delayed_events.py:
```diff
@@ -0,0 +1,95 @@
+    SEND = "send"
+
+
+# TODO: Needs unit testing
```
e41b5a1d74 adds a unit test module for the delayed event servlets.

For Complement tests, see https://github.com/matrix-org/complement/pull/734

AndrewFerr replied:
fb048335bb adds some more tests.

I'm unsure of a way to make Synapse unit tests stop/restart its homeserver, so currently only the Complement tests cover the case of delayed events that timed out while the server was down.

AndrewFerr commented:
Despite the failures in Sytest, it passes when I run it locally.

anoadragon453 reviewed (commented):
Thanks for all the time you've spent (and patiently waited) on this. This now looks good to me, and the added tests help.

I see you have a matching Complement branch here: https://github.com/matrix-org/complement/pull/734. I assume you would like to hold off on merging this PR until those tests pass?

Finally, once this is merged and done, if you have some time to document how the whole mechanism works (however detailed or briefly) in https://element-hq.github.io/synapse/latest/development/internal_documentation/index.html, that would help future development efforts.

AndrewFerr commented:
> I see you have a matching Complement branch here: [matrix-org/complement#734](https://github.com/matrix-org/complement/pull/734). I assume you would like to hold off on merging this PR until those tests pass?

Yes, as long as it's permissible to merge Complement tests before any implementations are merged.



> Finally, once this is merged and done, if you have some time to document how the whole mechanism works (however detailed or briefly) in https://element-hq.github.io/synapse/latest/development/internal_documentation/index.html, that would help future development efforts.

Noted!

anoadragon453 commented:
Typically we'll merge them at the same time. However, yes, technically the Complement tests can be merged beforehand, as they won't be enabled on `develop` until this PR is merged.

anoadragon453 commented:
(Complement CI is expected to fail until https://github.com/matrix-org/complement/pull/734 is merged.)

anoadragon453 commented:
https://github.com/matrix-org/complement/pull/734 has been merged. I've re-run the failed Complement jobs: https://github.com/element-hq/synapse/actions/runs/10891540744?pr=17326

anoadragon453 reviewed (approved):
LGTM, assuming the Complement tests are happy.

AndrewFerr commented:
> assuming the Complement tests are happy

They aren't :upside_down_face:  Looking into it now.  The failures reproduce locally, so I should have a fix ready very soon.

anoadragon453 commented:
@AndrewFerr looks like there's complement failures on the delayed event tests: https://github.com/element-hq/synapse/actions/runs/10891540744/job/30325234632?pr=17326

AndrewFerr commented:
Looks like the Complement failures are false-positives when running in Postgres.  I'm updating the Complement tests to account for that.

AndrewFerr commented:
The Postgres tests should be fixed by https://github.com/matrix-org/complement/pull/737.  They pass when I run them locally, in either monolith or worker mode.

AndrewFerr commented:
Alright, I added an extra commit to satisfy what the Complement test was failing on, and now we don't have to worry about merging https://github.com/matrix-org/complement/pull/737 first.

The test expects `GET /delayed_events` to return events in the order they'll be sent, but no such order was guaranteed when using Postgres.  [`dfde3c2`](https://github.com/element-hq/synapse/pull/17326/commits/dfde3c22db2490b8aa6b75813368d006ba32bf0c) applies the ordering the test expects.

Strictly speaking, the test doesn't have to cover ordering on `GET /delayed_events` because the MSC doesn't (yet) define an ordering, so I wrote https://github.com/matrix-org/complement/pull/737/commits/279d1b3724cef49ed8f8357900ef7ddb59bce73c to remove that part of the test.

The other test failure was simply due to leftover data after the first test failed, so I wrote https://github.com/matrix-org/complement/pull/737/commits/6f86880ab2ae6f394146d3b2c15721b2eef15745 to make tests do some teardown to prevent unrelated failures later on.

Note that all tests will continue to pass with https://github.com/matrix-org/complement/pull/737, so it can still be merged safely, but now it is optional to do so :slightly_smiling_face:
